Bugfix/localization 926

mandrade2 opened the pull request:
In this PR issue #253 is resolved. Whenever a submission is being saved, a stored procedure checks that with a `SELECT ... FOR UPDATE` row lock that the submission to be saved is not there yet. This prevents duplicate submission being saved. Making rows unique setting a multi-column index or unique_together in Django would've been preferred, but the older duplicate submissions needed to be there for consistency.

Travis jobs where configured to run with Mysql, so the migration and testing job succeed.

All database defaults where set to have a correct character set and collation.

iafan commented on pootle/apps/pootle_store/views.py:
```diff
@@ -598,7 +600,12 @@ def submit(request, unit):
                     old_value = str(suggestion.target_f)
-                sub = Submission(
+
+                if current_time.microsecond > 0:
```
Why do we need to manipulate with the submission time here?

mandrade2 replied:
It shouldn't go here. It slipped from an alternate solution where I would add milliseconds support on MySQL. Deleting it

iafan commented on Makefile:
```diff
@@ -52,6 +52,9 @@ lint-css:
 	cd ${JS_DIR} \
 	&& npm run stylelint
 
+reformat: 
```
This specific change is not related to this PR, it seems. It would be more clear to add it as a separate PR.

mandrade2 replied:
Sound's good. I'll split it then

iafan commented on .gitignore:
```diff
@@ -46,6 +46,7 @@ manifest.json
 /pootle/locale/*/LC_MESSAGES/
 /pootle/log/
 /pootle/static/translations/
+/pootle/invoices/
```
This seems to be related to how you set things up, not to the project code itself. Let's remove this rule from the PR; you can always add personal ignore rules to .git/info/exclude file in your local checkout.

mandrade2 replied:
Didn't knew there was a local exclude file. Great then

mandrade2 replied:
👌

iafan commented on pootle/apps/pootle_store/migrations/0006_auto_20200610_2137.py:
```diff
@@ -0,0 +1,44 @@
+            sql=[
+                (
+                    """
+                CREATE PROCEDURE `insert_unique_submission` ( IN 
```
I'd rename the procedure to `insert_submission`. The fact that it's unique shouldn't slip into the name (we will — hopefully! — never insert non-unique submission records).

mandrade2 replied:
Agree

iafan commented on pootle/apps/pootle_store/migrations/0006_auto_20200610_2137.py:
```diff
@@ -0,0 +1,44 @@
+                BEGIN 	
+                    DECLARE total_rows INT DEFAULT 0;      	
+
+                    SELECT COUNT(unit_id) INTO total_rows FROM `pootle_app_submission` WHERE `pootle_app_submission`.`creation_time` = creation_time AND `pootle_app_submission`.`field` = field_arg AND `pootle_app_submission`.`submitter_id` = submitter_id         AND `pootle_app_submission`.`type` = type_arg AND `pootle_app_submission`.`unit_id` = unit_id FOR UPDATE;
```
Please reformat the query to avoid overly long lines. You can insert line breaks before `FROM`, `WHERE`, `AND`, etc.

mandrade2 replied:
👌

iafan commented on pootle/apps/pootle_store/migrations/0006_auto_20200610_2137.py:
```diff
@@ -0,0 +1,44 @@
+                    SELECT COUNT(unit_id) INTO total_rows FROM `pootle_app_submission` WHERE `pootle_app_submission`.`creation_time` = creation_time AND `pootle_app_submission`.`field` = field_arg AND `pootle_app_submission`.`submitter_id` = submitter_id         AND `pootle_app_submission`.`type` = type_arg AND `pootle_app_submission`.`unit_id` = unit_id FOR UPDATE;
+
+                    IF total_rows = 0 THEN 
+                        INSERT INTO `pootle_app_submission` ( `pootle_app_submission`.`creation_time`, `pootle_app_submission`.`field`, `pootle_app_submission`.`submitter_id`, `pootle_app_submission`.`type`, `pootle_app_submission`.`unit_id`, `pootle_app_submission`.`store_id`, `pootle_app_submission`.`translation_project_id`, `pootle_app_submission`.`old_value`, `pootle_app_submission`.`new_valu ...
```
Same as above (format the query to be more readable)

iafan commented on pootle/apps/pootle_store/views.py:
```diff
@@ -611,7 +618,30 @@ def submit(request, unit):
                     similarity=form.cleaned_data["similarity"],
                     mt_similarity=form.cleaned_data["mt_similarity"],
                 )
-                sub.save()
```
Previously, we would create a `Submission()` object, then run a `save()` method on it that would store it in the database. Now we're creating this object but not doing anything with it, since the insertion happens inline in the code below.

I suggest moving the submission logic to the `Submission.save()` method, replacing the old code that should no longer be used, anyway, as I understand it. This will be cleaner, and will guarantee that submissions are always saved with the new method.

mandrade2 replied:
I wasn't sure if we needed the new logic on every submission save or only on this view. I'll move it then, definitely cleaner

iafan commented on pootle/settings/91-travis.conf:
```diff
@@ -16,28 +16,31 @@ if os.environ.get("TRAVIS"):
             'HOST': '',
             'PORT': '',
             'ATOMIC_REQUESTS': True,
+            'CHARSET':'utf8mb4',
```
Do we need to override CHARSET and COLLATION now that they are in `90-local.conf.template` file? (90-local.conf file will be loaded before 91-travis.conf)

mandrade2 replied:
But is Travis loading `90-local.conf.template` on the build environment? It seems to me that It should not run it since it has the `.template` extension

mandrade2 replied:
Any input on this Igor? As I stated above It's not clear to me that travis will use the configuration in `90-local.conf.template`

iafan commented on pootle/settings/91-travis.conf:
```diff
@@ -16,28 +16,31 @@ if os.environ.get("TRAVIS"):
-                'NAME': '',
-                'CHARSET': 'utf8',
+                'NAME': 'test_zing',
+                'CHARSET':'utf8mb4',
```
Same question as above.

iafan commented on pootle/settings/91-travis.conf:
```diff
@@ -16,28 +16,31 @@ if os.environ.get("TRAVIS"):
-    if DATABASE_BACKEND.startswith("mysql"):
-        DATABASES['default']['ENGINE'] = 'django.db.backends.mysql'
-        DATABASES['default']['NAME'] = 'zing'
+    if DATABASE_BACKEND.startswith("sqlite"):
```
Can you please explain this change? I think with this PR we're moving away from sqlite.

mandrade2 replied:
It's to maintain sqlite support for user's that are on it. Thought I should keep it for compatibility

iafan commented on tests/conftest.py:
```diff
@@ -123,7 +116,7 @@ def pytest_addoption(parser):
 def should_generate_snapshots(pytestconfig):
     """Indicates whether snapshots have been instructed to be generated."""
-    return bool(pytestconfig.getoption("generate_snapshots"))
+    return True
```
Why do we generate snapshots unconditionally now?

mandrade2 replied:
Leftover from some tests I was running. Fixing it

iafan commented on tests/data/snapshots/test_browse/admin/_language0_.context.json:
```diff
@@ -9,11 +9,11 @@
-          "mtime": 1578661829,
-          "source": "Translated Source /language0/project0/subdir0/sto\u2026",
+          "mtime": 1591122676,
+          "source": "Translated Source /language0/project0/store0.po 2\u2026",
```
Here and below, I'm concerned about these snapshot changes to the source string. They don't seem to be related to the changes this PR is supposed to introduce. We need to understand why this property is reported differently now.

mandrade2 replied:
Don't know why it happens. Tests wouldn't pass if I kept the old snapshots. Might be because of my environment, but before the changes I introduced the tests would run smoothly.

mandrade2 replied:
The thing is that when I run the tests without generating a new dump (not snapshots) I get this error on all tests when setting up the fixtures.
![Screen Shot 2020-07-09 at 1 03 09 PM](https://user-images.githubusercontent.com/16262338/87069185-c03aba80-c1e4-11ea-95be-6babc5c3ef9a.png)
Then, when I generate a new dump I can get past the fixtures setup but other errors show up. 

When I called the submission procedure outside the save method, generating a new dump and snapshots would fix all tests. However, there's that weird modification of source strings.

So the problem here would be that the original `data_dump.json` is having troubles when initialized on MySQL. Apparently because of uniqueness constraints. On the other hand when I create a new dump some of this problems dissapear.

mandrade2 replied:
There was no reason to modify snapshots, fixed.

mandrade2 replied:
@iafan It turns out I could not get over the uniqueness constraints colliding from the dump without generating a new dump, so I uploaded a new one. It contains some changes on ids, timestamps and ordering. 

The new dump would not pass all tests without regenerating snapshots, so I did regenerate them. The new snapshots have some weird source changes as you pointed out. I've looked back at commit history and found [a commit](https://github.com/evernote/zing/commit/81f4b9391b4d4af896a98b7f51024819b5093e9e) that had a similar impact on snapshots. The last modification of urls to be tested occurs [here]( https://github.com/evernote/zing/commit/107746b7f1e815c587c18e5ccbb7b8513c297a01), so I discard the source modifications are due to different urls being tested and just having out of date snapshots.

mandrade2 commented:
I fixed the remaining issues regarding tests. It turns out that I had to modify the submission save method to call for a reload from the database, to avoid Django raising an error because of garbage collecting an unsaved instance. Also added to the stored procedure arguments for a suggestion_id and quality_check_id, I had forgotten them.

I rolled back all the snapshot changes and fixed two tests that kept failing. The `_test_user_purging` test was failing because when asserting that two datetimes where equal, the dates from MySQL were coming back from the db without microseconds. My solution was to strip microseconds of the object that we test against. The other test was `test_timeline_view_unit_with_creation`. It failed because the submission unit was saved twice, which in turn called twice a method from the Unit save method called `add_initial_submission`. Plus, there was another call to `add_initial_submission` on the test itself. I removed one call to the unit save method and also the extra `add_initial_submission` call. I'm not sure if this changes too much the tested case on the test. You can see on both tests that it was expected for them to behave different when being tested on MySQL.
